unleash-yggdrasil/src/strategy_parsing.rs

```diff
@@ -46,6 +46,23 @@ pub fn normalized_hash(
     murmur3_32(&mut reader, seed).map(|hash_result| hash_result % modulus + 1)
 }
 
+fn drain<const N: usize>(node: Pairs<Rule>) -> [Pair<Rule>; N] {
+    drain_partial(node).0
+}
+
+fn drain_partial<const N: usize>(mut node: Pairs<Rule>) -> ([Pair<Rule>; N], Pairs<Rule>) {
```

Comment:
I don't even care if you understand this, this is just so damn cool I'm gonna talk about it.

This is generic over a compile time constant usize: the compiler will build us one implementation for every value between 0 and usize max that we use. That sounds like a lot but we only use this for 1, 2 and 3 so it's okay. This function yields a compile time checked array of size N (the constant we're generic over). 

What that allows us to do is get exactly N children from our node and enforce that at compile time.

Because Rust is awesome, the compiler will infer a ton of that information for us.

So this code:

```
let [context_getter, ordinal_operation, number] = drain(node);
```

Is just sugar for this (pseudo-code):

```
let children: [Rule; 3]= drain(node); //we remove exactly 3 nodes and unwrap them, this compile time checked
let [first, second, third] =  children;
```

So we're effectively delegating our compile time checking of the Yggdrasil DSL to the Rust compiler!

Reply:
Lol, `I don't even care if you understand this`. Kinda harsh? I love the result here though. :+1:

Reply:
Haha, I'm just super excited and I want the opportunity to talk about how cool Rust is!